File: tools-perf-hypack-2013/tools-perf-codes/Fortran/Tools_MM_Mult_Master_Slave.f
c
c*******************************************************************
c
c		C-DAC Tech Workshop : hyPACK-2013
c                           October 15-18, 2013
c
c*******************************************************************
c
c                Example 5 (Tools_MM_Mult_Master_Slave.f)
c
c  Objective           : Matrix_Matrix_Multiplication
c                        (Self_Scheduling algorithm or master-slave program)
c
c  Description         : In this is the worker(rank!=0) which will be executed
c                        by the rest of the processors.
c                        The workers does the computations on the data
c                        distributed by the master(rank==0) & indicates to the 
c                        master for which tow it has done the computation.
c  Input               : Process 0 (master) reads MatrixA and MatrixB
c
c  Output              : Process 0 prints the result of Matrix-Matrix
c                        Multiplication
c
c  Necessary Condition : Number of Processes should be less than
c                        or equal to 8
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c*******************************************************************

      program main
      include 'mpif.h'

      integer MAX_ROWS, MAX_COLS
      parameter (MAX_ROWS = 100, MAX_COLS = 100)
c

      integer MAX_ROWS, MAX_COLS, Dest, RowtoSend, Source,row
      integer NoofRowsA, NoofColsA
      integer NoofRowsB, NoofColsB

      double precision Matrix_A(MAX_ROWS,MAX_COLS), 
     $	               Matrix_B(MAX_ROWS,MAX_COLS), 
     $                 ResultMatrix(MAX_ROWS,MAX_COLS)
      double precision RowA(MAX_COLS),  ResultRow(MAX_COLS)

      real*8 Starttime, Endtime, Time
c
      integer MyRank, Root, Numprocs
      integer status(MPI_STATUS_SIZE)
      integer i, j, k, iproc, ierror
      integer rowtype, anstype, donetype
      integer Source_tag, Destination_tag,Seed,k
c

      call MPI_INIT(ierror)
      call MPI_COMM_RANK(MPI_COMM_WORLD, MyRank, ierror )
      call MPI_COMM_SIZE(MPI_COMM_WORLD, Numprocs, ierror )
c
      rowtype  = 1
      anstype  = 2
      donetype = 0
      Root     = 0
      Seed     = 760013

c     Initialize Matrix_A and Matrix_B

      if(MyRank .eq. Root) then
      Starttime = MPI_WTIME()
			 print*, " enter the rows and columns of Matrix_A"
          read(*,*) NoofRowsA 
          read(*,*) NoofColsA 
         do i = 1,NoofRowsA
         do j = 1,NoofColsA
             Matrix_A(i,j) = ran(Seed)*10 
         enddo
         enddo
			 
			 print*, " enter the rows and columns of Matrix_B"
          read(*,*) NoofRowsB 
          read(*,*) NoofColsB  
         do i = 1,NoofRowsB
         do j = 1,NoofColsB
            Matrix_B(i,j) = ran(Seed)*10
         enddo
         enddo

	call MPI_Barrier(MPI_COMM_WORLD, ierror)
	call MPI_Bcast(NoofColsA, 1, MPI_INTEGER, Root,
     $ 	               MPI_COMM_WORLD,ierror)

	call MPI_Bcast(NoofRowsB, 1, MPI_INTEGER, Root,
     $  	       MPI_COMM_WORLD,ierror)

	if(NoofColsA .ne. NoofRowsB) then
  	 print*,"Incompatible dimensions of Matrices for Mutlipication"
   	 goto 100
        endif

	call MPI_Bcast(NoofColsB, 1, MPI_INTEGER, Root, 
     $ 	               MPI_COMM_WORLD,ierror)

c        send Matrix_B to each other process
	 do 86 i = 1, NoofColsB
          call MPI_BCAST(Matrix_B(1,i),NoofRowsB,MPI_DOUBLE_PRECISION, 
     $          Root,MPI_COMM_WORLD,ierror)
  86	continue
c
c        send a row of Matrix_A to each other process
c
	 RowtoSend = 0
         do 40 iproc = 1, Numprocs-1
            do 30 j = 1,NoofColsA
               RowA(j) = Matrix_A(iproc,j)
  30         continue
c
            Dest = iproc
               call MPI_SEND(RowA, NoofColsA, MPI_DOUBLE_PRECISION,
     $              Dest, RowtoSend+1, MPI_COMM_WORLD, ierror)
               RowtoSend = RowtoSend+1
  40      continue
c
         do 70 i = 1, NoofRowsA
          call MPI_RECV( ResultRow, NoofColsA, MPI_DOUBLE_PRECISION,
     $      MPI_ANY_SOURCE,MPI_ANY_TAG,MPI_COMM_WORLD,status,ierror)
c
             Dest = status(MPI_SOURCE)
             Source_tag = status(MPI_TAG)
c
             do k = 1, NoofColsA 
	             ResultMatrix(Source_tag,k) =  ResultRow(k)
             end do
c
               if(RowtoSend .lt. NoofRowsA) then
               do 50 j = 1,NoofColsA
                  RowA(j) = Matrix_A(RowtoSend+1,j)
  50            continue
c
	     Destination_tag = RowtoSend + 1
             call MPI_SEND(RowA,NoofColsA,MPI_DOUBLE_PRECISION,Dest,
     $                       Destination_tag, MPI_COMM_WORLD,ierror)
               
               RowtoSend     = RowtoSend+1
            else
               RowA(1) = 1.0d0
               call MPI_SEND(RowA, 1, MPI_DOUBLE_PRECISION, 
     $                     Dest, donetype, MPI_COMM_WORLD, ierror)
            endif
	Endtime = MPI_Wtime()
	Time    = EndTime - Starttime
  70      continue
        write(6,75)((ResultMatrix(k,j), j=1, NoofColsA), k=1, NoofRowsA)
	print*, "  Time = ", Time
       else 

        call MPI_Barrier(MPI_COMM_WORLD, ierror)
        call MPI_Bcast (NoofColsA, 1, MPI_INTEGER, Root,
     $           MPI_COMM_WORLD, ierror)

        call MPI_Bcast (NoofRowsB, 1, MPI_INTEGER, Root,
     $          MPI_COMM_WORLD, ierror)

        if(NoofColsA .ne. NoofRowsB) then
          goto 100
        endif

        call MPI_Bcast (NoofColsB, 1, MPI_INTEGER, Root,
     $          MPI_COMM_WORLD, ierror)

C
C       slave receive B, then compute rows of C until done message.
C
        do 85 i = 1, NoofColsB

         call MPI_BCAST(Matrix_B(1,i),NoofColsA,MPI_DOUBLE_PRECISION,
     $            Root,MPI_COMM_WORLD,ierror)
  85      continue 
        Source = Root
  90      call MPI_RECV(RowA,NoofColsA,MPI_DOUBLE_PRECISION,Source,
     $        MPI_ANY_TAG, MPI_COMM_WORLD, status, ierror)

c
        if (status(MPI_TAG) .eq. 0) then
            go to 200
        else
           row = status(MPI_TAG)
            do 110 i = 1, NoofColsB
               ResultRow(i) = 0.0
               do 95 j = 1, NoofColsA
                 ResultRow(i) = ResultRow(i) + RowA(j)*Matrix_B(j,i)
  95           continue
  110       continue
c
            call MPI_SEND(ResultRow, NoofColsB, MPI_DOUBLE_PRECISION,
     $                    Root, row, MPI_COMM_WORLD, ierror)
            go to 90
         endif
  200   continue    
        endif   
c        if(MyRank .eq. Root) then 
c        write(6,75)((ResultMatrix(i,j), j=1, NoofColsA), i=1, NoofRowsA)
c	print*, "  Time = ", Time
c        endif

  100  	call MPI_FINALIZE(ierror)
  75	   format(8(2x,f8.3))
      stop
      end

File: tools-perf-hypack-2013/tools-perf-codes/c/Tools_MM_Mult_Master_Slave.c
/*
*******************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

 Example 5   :(Tools_MM_Mult_Master_Slave.c)	

 Objective   : Matrix_Matrix_Multiplication
              (Self_Scheduling algorithm or Master Slave)

 Description : In this the processor other than rank 0 will do 
	       the computations on the data distributed by the processor 
               with rank 0 and indicates to the root for which row it has
               done the computation.

 Input       : Read Matrix 1 & Matrix 2.

 Output      : Result of Matrix-Matrix Multiplication on processor 0.

 Necessary Condition : Number of Processes should be
		       less than or equal to 8

   Created          : August-2013

   E-mail           : hpcfte@cdac.in     

*******************************************************************
*/

#include <stdio.h>
#include <math.h>
#include "mpi.h"
#define SIZE 16


main(int argc, char** argv) 
{

 int	     Numprocs, MyRank;
 int		  NoofRowsA, NoofColsA,	NoofRowsB, NoofColsB;
 int		  Source, Destination, Root=0;
 int	     Source_tag, Destination_tag;
 int 	     flag,irow, icol, iproc, index;
 int	     RowtoSend = 0, RowReceived = 0;
 float	     **Matrix_A, **Matrix_B, **ResultMatrix;
 float	     *BufferB, *RowA, *ResultRow;
 MPI_Status status;     

 /* ........MPI Initialisation .......*/
	MPI_Init(&argc, &argv); 
  	MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);
  	MPI_Comm_size(MPI_COMM_WORLD, &Numprocs);

 /* Reading Input */
if(MyRank==0)
{
     NoofRowsA=SIZE;
     NoofColsA=SIZE;
     Matrix_A = (float **) malloc (NoofRowsA * sizeof(float *));
     for (irow = 0; irow < NoofRowsA; irow++){
	Matrix_A[irow] = (float *) malloc(NoofColsA * sizeof(float));
     		for (icol = 0; icol < NoofColsA; icol++)
       		Matrix_A[irow][icol]= rand(10);
     		}

     NoofRowsB=SIZE;
     NoofColsB=SIZE;
     Matrix_B = (float **) malloc (NoofRowsB * sizeof(float *));
     for(irow = 0; irow < NoofRowsB; irow++){
        Matrix_B[irow] = (float *) malloc(NoofColsB * sizeof(float *));
       		for(icol = 0; icol < NoofColsB; icol++)
     		Matrix_B[irow][icol]=rand(10);
    		}

     MPI_Barrier(MPI_COMM_WORLD);
     MPI_Bcast (&NoofColsA, 1, MPI_INT, 0, MPI_COMM_WORLD);
     MPI_Bcast (&NoofRowsB, 1, MPI_INT, 0, MPI_COMM_WORLD);

   if(NoofColsA != NoofRowsB){
    MPI_Finalize();
    printf("Incompatible dimensions of Matrices for Mutlipication\n");
    exit(0);
 }

   MPI_Bcast (&NoofColsB, 1, MPI_INT, 0, MPI_COMM_WORLD);

 /* .......Convert Matrix B into 1-D BufferB and Broadcast to Workers.....*/
 BufferB = (float *) malloc(NoofRowsB*NoofColsB*sizeof(float));
 index  = 0;
 for(irow = 0; irow<NoofRowsB; irow++)
   for(icol=0; icol<NoofColsB; icol++)
   BufferB[index++] = Matrix_B[irow][icol];
   MPI_Bcast (BufferB,NoofRowsB*NoofColsB,MPI_FLOAT,0,MPI_COMM_WORLD);

 /* .......Allocate memory for output and a row of a Matrix A......*/ 
   RowA = (float *)malloc(NoofColsA*sizeof(float));
   ResultMatrix = (float **)malloc(NoofRowsA*sizeof(float *)); 
   for(irow=0; irow<NoofRowsA; irow++)
     ResultMatrix[irow] = (float *)malloc(NoofColsB*sizeof(float));

   /* First Send At least one row to each processors */
    RowtoSend = 0;
    for(iproc = 1 ; iproc < Numprocs ; iproc++) {
	for(icol=0; icol<NoofColsA; icol++)
	RowA[icol] = Matrix_A[RowtoSend][icol];
   MPI_Send(RowA, NoofColsA, MPI_FLOAT, iproc, RowtoSend+1, MPI_COMM_WORLD);
	RowtoSend++;
	}

   ResultRow = (float *) malloc(NoofColsB*sizeof(float));
	for(irow = 0 ; irow < NoofRowsA ; irow++) {
	MPI_Recv(ResultRow, NoofColsB, MPI_FLOAT, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
	Destination = status.MPI_SOURCE;
	Source_tag  = status.MPI_TAG;
		 
	memcpy(ResultMatrix[Source_tag], ResultRow, NoofColsB*sizeof(float));

 /*For the  remaining rows(if any) send the data to whichever processor that finishes the earlier computations first */

   if( RowtoSend < NoofRowsA ) {
	 memcpy(RowA, Matrix_A[RowtoSend], NoofColsA*sizeof(float));
 	 Destination_tag = RowtoSend+1;
   	 MPI_Send(RowA, NoofColsA, MPI_FLOAT, Destination, Destination_tag, MPI_COMM_WORLD);
	 RowtoSend++;
    }

	}

 for(iproc = 1 ; iproc < Numprocs ; iproc++) {
   MPI_Send(RowA, NoofColsA, MPI_FLOAT, iproc, 0, MPI_COMM_WORLD);
}
}
if(MyRank!=0)
{

MPI_Barrier(MPI_COMM_WORLD);

      MPI_Bcast (&NoofColsA, 1, MPI_INT, 0, MPI_COMM_WORLD);
      MPI_Bcast (&NoofRowsB, 1, MPI_INT, 0, MPI_COMM_WORLD);
 if(NoofColsA != NoofRowsB){
              MPI_Finalize();
              exit(0);
      }
       MPI_Bcast (&NoofColsB, 1, MPI_INT, 0, MPI_COMM_WORLD);

  BufferB = (float *)malloc(NoofRowsB*NoofColsB*sizeof(float));
       MPI_Bcast (BufferB, NoofRowsB*NoofColsB, MPI_FLOAT, 0, MPI_COMM_WORLD);

  ResultRow = (float *)malloc(NoofColsB*sizeof(float));
       RowA        = (float *)malloc(NoofColsA*sizeof(float));

  for(;;) {
          MPI_Recv(RowA, NoofColsA, MPI_FLOAT, Root, MPI_ANY_TAG,
                                       MPI_COMM_WORLD, &status);

     flag = status.MPI_TAG - 1;
          if(flag == -1)
                       break;

               Destination = Root;
               Destination_tag = flag;

               for(icol = 0; icol< NoofColsB; icol++) {
                       ResultRow[icol] = 0;
                       for(irow=0; irow< NoofRowsB; irow++) {
                               index = irow*NoofColsB + icol;
                          ResultRow[icol] += BufferB[index] * RowA[irow];

     
                        }
                }
                MPI_Send(ResultRow, NoofColsB, MPI_FLOAT, 0, Destination_tag,MPI_COMM_WORLD);
   }
}

if(MyRank==0)
{
	
  printf("Result is ...\n");
	for(irow = 0 ; irow < NoofRowsA; irow++) {
		for( icol= 0 ;  icol< NoofColsB; icol++) 
			printf(" %3.3f ",ResultMatrix[irow][icol]); 
	   printf("\n");
	
}
}
   MPI_Finalize(); 
}
